go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+)
+
+// UsePvl says whether to use PVL for verifying proofs.
+const UsePvl = false
+
+// PvlSupportedVersion is which version of PVL is supported by this client.
+const PvlSupportedVersion int = 1
+
+type PvlScriptState struct {
+	PC           int32
+	Service      keybase1.ProofType
+	Vars         PvlScriptVariables
+	ActiveString string
+	FetchURL     string
+	HasFetched   bool
+	// nil until fetched
+	FetchResult *PvlFetchResult
+}
+
+type PvlScriptVariables struct {
+	UsernameService string
+	UsernameKeybase string
+	Sig             []byte
+	SigIDMedium     string
+	SigIDShort      string
+	Hostname        string
+}
+
+type PvlFetchResult struct {
+	Mode PvlMode
+	// One of these 3 must be filled.
+	String string
+	HTML   *goquery.Document
+	JSON   *jsonw.Wrapper
+}
+
+type PvlMode string
+
+const (
+	PvlModeJSON   PvlMode = "json"
+	PvlModeHTML   PvlMode = "html"
+	PvlModeString PvlMode = "string"
+	PvlModeDNS    PvlMode = "dns"
+)
+
+const (
+	PvlAssertRegexMatch    = "assert_regex_match"
+	PvlAssertFindBase64    = "assert_find_base64"
+	PvlWhitespaceNormalize = "whitespace_normalize"
+	PvlRegexCapture        = "regex_capture"
+	PvlFetch               = "fetch"
+	PvlSelectorJSON        = "selector_json"
+	PvlSelectorCSS         = "selector_css"
+	PvlTransformURL        = "transform_url"
+)
+
+type PvlStep func(GlobalContextLite, *jsonw.Wrapper, PvlScriptState) (PvlScriptState, ProofError)
+
+var PvlSteps = map[string]PvlStep{
+	PvlAssertRegexMatch:    pvlStepAssertRegexMatch,
+	PvlAssertFindBase64:    pvlStepAssertFindBase64,
+	PvlWhitespaceNormalize: pvlStepWhitespaceNormalize,
+	PvlRegexCapture:        pvlStepRegexCapture,
+	PvlFetch:               pvlStepFetch,
+	PvlSelectorJSON:        pvlStepSelectorJSON,
+	PvlSelectorCSS:         pvlStepSelectorCSS,
+	PvlTransformURL:        pvlStepTransformURL,
+}
+
+// Checkproof verifies one proof by running the pvl on the provided proof information.
+func CheckProof(g GlobalContextLite, pvl *jsonw.Wrapper, service keybase1.ProofType, link RemoteProofChainLink, h SigHint) ProofError {
+	if perr := pvlValidateChunk(pvl, service); perr != nil {
+		return perr
+	}
+
+	sigBody, sigID, err := OpenSig(link.GetArmoredSig())
+	if err != nil {
+		return NewProofError(keybase1.ProofStatus_BAD_SIGNATURE,
+			"Bad signature: %v", err)
+	}
+
+	scripts, perr := pvlChunkGetScripts(pvl, service)
+	if perr != nil {
+		return perr
+	}
+
+	newstate := func() PvlScriptState {
+		vars := PvlScriptVariables{
+			UsernameService: link.GetRemoteUsername(), // Blank for DNS-proofs
+			UsernameKeybase: link.GetUsername(),
+			Sig:             sigBody,
+			SigIDMedium:     sigID.ToMediumID(),
+			SigIDShort:      sigID.ToShortID(),
+			Hostname:        link.GetHostname(), // Blank for non-{DNS/Web} proofs
+		}
+
+		// Enforce prooftype-dependent variables.
+		webish := (service == keybase1.ProofType_DNS || service == keybase1.ProofType_GENERIC_WEB_SITE)
+		if webish {
+			vars.UsernameService = ""
```

Comment:
We might get less confusing errors if we use a sentinel here like `"<<<INVALID>>>"` or whatever, instead of the empty string. Otherwise somewhere down the line we're going to see `assertion failed, "foobar" != ""` and forget what that means.